### PR TITLE
feat: Implement domain scoping

### LINKED ISSUE #257: [FEATURE] Implement domain scoping
## Overview

A `domain` is an identifier that logically binds clients with providers, allowing multiple providers to be used simultaneously within a single application.

Domains were added to the spec in the following PR: https://github.com/open-feature/spec/pull/229

## Tasks

```[tasklist]
### Tasks
- [ ] [Requirement 1.1.3](https://openfeature.dev/specification/sections/flag-evaluation#requirement-113): The API MUST provide a function to bind a given provider to one or more clients using a domain. If the domain already has a bound provider, it is overwritten with the new mapping.
- [ ] [Requirement 1.1.6](https://openfeature.dev/specification/sections/flag-evaluation#requirement-116): The API MUST provide a function for creating a client which accepts `domain`
- [ ] [Requirement 1.2.2](https://openfeature.dev/specification/sections/flag-evaluation#requirement-122): The client interface MUST define a metadata member or accessor, containing an immutable domain field or accessor of type string, which corresponds to the domain value supplied during client creation.
- [ ] [Requirement 3.2.2.3](https://openfeature.dev/specification/sections/evaluation-context#conditional-requirement-3223): The API MUST have a method for setting evaluation context for a domain.
- [ ] [Requirement 3.2.2.4](https://openfeature.dev/specification/sections/evaluation-context#conditional-requirement-3224): The API MUST have a mechanism to manage evaluation context for an associated domain.
- [ ] Update the `named provider` section of the readme to `domain`. See the JS SDK readme for an example: https://github.com/open-feature/js-sdk/tree/main/packages/server#domains
``` 

## Reference implementations

Support was added in the JS SDK [here](https://github.com/open-feature/js-sdk/pull/805) and the Python SDK [here](https://github.com/open-feature/python-sdk/pull/271).

> [!NOTE]
> Some SDKs support `named clients` which predate `domains`. The `domain` term is intended to clarify intended behavior.

### PR DESCRIPTION
**Pull request by gunishmatta**
## This PR

- Adds Domain Scoping

### Related Issues

Fixes #257

### CODE REVIEW

File: openfeature/client.go [hunk 1]
```diff
@@ -57,10 +57,10 @@ type Client struct {
 	logger            func() logr.Logger
 }
 
-// NewClient returns a new Client. Name is a unique identifier for this client
+// NewClient returns a new Client. Domain is a unique identifier for this client
 func NewClient(name string) *Client {
```
Comment (beeme1mr) [inline openfeature/client.go:61]:
```suggestion
func NewClient(domain string) *Client {
```

File: openfeature/client.go [hunk 2]
```diff
@@ -57,10 +57,10 @@ type Client struct {
...
 	return &Client{
-		metadata:          ClientMetadata{name: name},
+		metadata:          ClientMetadata{domain: name},
```
Comment (beeme1mr) [inline openfeature/client.go:63]:
```suggestion
		metadata:          ClientMetadata{domain: domain},
```

File: openfeature/event_executor_test.go [hunk 3]
```diff
@@ -160,7 +160,7 @@ func TestEventHandler_Eventing(t *testing.T) {
 
 		// trigger event from provider implementation
 		eventingImpl.Invoke(Event{
-			ProviderName: eventingProvider.Metadata().Name,
+			ProviderName: eventingProvider.Metadata().Domain,
```
Comment (beeme1mr) [inline openfeature/event_executor_test.go:163]: This should remain provider name but the event could be extended to include a domain property.

File: openfeature/event_executor_test.go [hunk 4]
```diff
@@ -245,7 +245,7 @@ func TestEventHandler_clientAssociation(t *testing.T) {
 
 	// invoke default provider
 	eventingImpl.Invoke(Event{
-		ProviderName:         defaultProvider.Metadata().Name,
+		ProviderName:         defaultProvider.Metadata().Domain,
```
Comment (beeme1mr) [inline openfeature/event_executor_test.go:248]: Same comment as above. This should remain the provider name.

File: openfeature/client.go [hunk 5]
```diff
@@ -32,20 +32,20 @@ type IClient interface {
...
 	}
 }
 
-// Name returns the client's name
-func (cm ClientMetadata) Name() string {
-	return cm.name
+// Domain returns the client's domain
+func (cm ClientMetadata) Domain() string {
+	return cm.domain
 }
```
Comment (Kavindu-Dodan) [inline openfeature/client.go:49]: We should avoid this breaking change. 

Instead, keep the old method and make it deprecate (use comment `// Deprecated` [1]). And add the new method with `Domain() string` signature, returning the `cm.domain`

[1] - https://go.dev/wiki/Deprecated

File: openfeature/provider.go [hunk 6]
```diff
@@ -186,7 +186,7 @@ type InterfaceResolutionDetail struct {
 	ProviderResolutionDetail
 }
 
-// Metadata provides provider name
+// Metadata provides provider domain
 type Metadata struct {
-	Name string
+	Domain string
 }
```
Comment (Kavindu-Dodan) [inline openfeature/provider.go:192]: This should be revered :) Provider name is still valid and it's Provider's name (ex:- flagd provider https://github.com/open-feature/go-sdk-contrib/blob/providers/flagd/v0.1.22/providers/flagd/pkg/provider.go#L144-L148 )

File: openfeature/client.go [hunk 7]
```diff
@@ -38,22 +38,29 @@ type IClient interface {
 
 // ClientMetadata provides a client's metadata
 type ClientMetadata struct {
-	name string
+	domain string
+	name   string
```
Comment (Kavindu-Dodan) [inline openfeature/client.go:42]: we don't have to change the internal values yet, the getter `Domain()` mapping to the old internal field is enough. 

  ```suggestion
	name   string
```

File: openfeature/client.go [hunk 8]
```diff
@@ -38,22 +38,29 @@ type IClient interface {
...
 }
 
 // NewClientMetadata constructs ClientMetadata
 // Allows for simplified hook test cases while maintaining immutability
-func NewClientMetadata(name string) ClientMetadata {
+func NewClientMetadata(domain string) ClientMetadata {
 	return ClientMetadata{
-		name: name,
+		domain: domain,
```
Comment (Kavindu-Dodan) [inline openfeature/client.go:49]: please map to the same internal field `name`

File: pkg/openfeature/client.go [hunk 9]
```diff
@@ -29,7 +29,7 @@ func NewClientMetadata(name string) ClientMetadata {
 // Deprecated: use github.com/open-feature/go-sdk/openfeature.Client, instead.
 type Client = openfeature.Client
 
-// NewClient returns a new Client. Name is a unique identifier for this client
+// NewClient returns a new Client. Domain is a unique identifier for this client
```
Comment (Kavindu-Dodan) [inline pkg/openfeature/client.go:32]: we can skip this change as this is in a deprecated module

**Comment by Kavindu-Dodan** [comment]
@gunishmatta really nice work 🙌  could you please check the minor changes I proposed? They should be easy to complete and once done I think this is ready

**Comment by gunishmatta** [comment]
Hi @Kavindu-Dodan  I am done with the changes, please merge if looks good.

**Review by Kavindu-Dodan — APPROVED** [review]
Nice work @gunishmatta, looks goot to me :)

**Comment by toddbaert** [comment]
I'll review this tomorrow. Thanks!

**Review by toddbaert — APPROVED** [review]
Sorry for the delay @gunishmatta . Thanks a lot for the changes!

**Pull request merged**